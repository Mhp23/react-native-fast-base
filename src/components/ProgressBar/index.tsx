import React from 'react';
import {colorSelector} from '../../utils';
import {useStyle, useTheme} from '../../hooks';
import {useRM} from 'react-native-full-responsive';
import {Animated, StyleSheet, View} from 'react-native';
import {DefaultSizes, PrimaryColors, ProgressProps} from '../../core';

const MIN_VALUE = 0;
const MAX_VALUE = 100;

const defaultProps: Omit<ProgressProps, 'width' | 'value'> = {
  height: 'md',
  useNativeDriver: true,
};

const Progress: React.FC<ProgressProps> = ({
  style,
  value,
  isRTL,
  width,
  height,
  background,
  progressColor,
  useNativeDriver,
}) => {
  const {rs} = useRM();

  const {colors} = useTheme();

  const animatedWidth = React.useRef(new Animated.Value(value));

  const progressWidth = React.useMemo(() => {
    if (typeof width === 'number') {
      return rs(width);
    } else {
      throw new Error('width of progress bar should be numeric value');
    }
  }, [rs, width]);

  const containerStyle = useStyle(() => {
    const progressHeight = rs(
      typeof height === 'number' ? height : DefaultSizes[height] - 2,
    );
    const backgroundColor = colorSelector(background) || colors?.flat;
    return StyleSheet.flatten([
      {
        backgroundColor,
        overflow: 'hidden',
        width: progressWidth,
        height: progressHeight,
        borderRadius: progressWidth / 2,
      },
      style,
    ]);
  }, [rs, height, background, colors?.flat, progressWidth, style]);

  const progressBarStyle = useStyle(() => {
    const backgroundColor =
      colorSelector(progressColor) || PrimaryColors.primary;
    return {
      width: '100%',
      height: '100%',
      backgroundColor,
    };
  }, [progressColor]);

  React.useEffect(() => {
    /**
     * set progress value between 0 to 100 if was been outside of this range
     */
    const toValue = Math.min(Math.max(value, MIN_VALUE), MAX_VALUE);

    Animated.spring(animatedWidth.current, {
      toValue,
      useNativeDriver,
    }).start();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);

  return (
    <View style={containerStyle}>
      <Animated.View
        testID="FAST_BASE_PROGRESS_BAR"
        style={[
          progressBarStyle,
          {
            transform: [
              {
                translateX: animatedWidth.current.interpolate({
                  inputRange: [0, 100],
                  outputRange: [
                    (isRTL ? progressWidth : -1 * progressWidth) / 2,
                    0,
                  ],
                }),
              },
              {
                scaleX: animatedWidth.current.interpolate({
                  inputRange: [0, 100],
                  outputRange: [0, 1],
                }),
              },
            ],
          },
        ]}
      />
    </View>
  );
};

Progress.defaultProps = defaultProps;

export default Progress;
