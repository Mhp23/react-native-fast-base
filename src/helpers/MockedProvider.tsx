import React from 'react';
import {DefaultTheme, ThemeContext, type ThemeContextProps} from '../core';

const MockedProvider: React.FC = ({children}) => {
  return (
    <ThemeContext.Provider
      value={{
        ...({} as ThemeContextProps),
        ...DefaultTheme.DefaultTheme,
      }}>
      {children}
    </ThemeContext.Provider>
  );
};

export default MockedProvider;
