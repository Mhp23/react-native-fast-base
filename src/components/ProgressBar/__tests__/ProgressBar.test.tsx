import React from 'react';
import ProgressBar from '..';
import MockedProvider from '../../../helpers/MockedProvider';
import {render, cleanup} from '@testing-library/react-native';

afterEach(cleanup);

describe('ProgressBar component', () => {
  it('should render correctly', () => {
    const rendered = render(
      <MockedProvider>
        <ProgressBar width={100} value={0} />
      </MockedProvider>,
    );
    expect(rendered).toBeTruthy();
    expect(rendered).toMatchSnapshot();
  });
});
