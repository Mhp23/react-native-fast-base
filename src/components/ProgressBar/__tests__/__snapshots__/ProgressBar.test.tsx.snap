// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar component should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ECF0F1",
      "borderRadius": 104.16666666666667,
      "height": 20.833333333333336,
      "overflow": "hidden",
      "width": 208.33333333333334,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "#3b82f6",
        "height": "100%",
        "transform": Array [
          Object {
            "translateX": -104.16666666666667,
          },
          Object {
            "scaleX": 0,
          },
        ],
        "width": "100%",
      }
    }
    testID="FAST_BASE_PROGRESS_BAR"
  />
</View>
`;
